e2e/tests/upgrades/genesis_test.go

```diff
@@ -0,0 +1,159 @@
+	})
+
+	t.Run("start relayer", func(t *testing.T) {
+		s.StartRelayer(relayer)
+	})
+
+	t.Run("packets are relayed", func(t *testing.T) {
+		s.AssertPacketRelayed(ctx, chainA, channelA.PortID, channelA.ChannelID, 1)
+
+		actualBalance, err := chainB.GetBalance(ctx, chainBAddress, chainBIBCToken.IBCDenom())
+		s.Require().NoError(err)
+
+		expected := testvalues.IBCTransferAmount
+		s.Require().Equal(expected, actualBalance)
+	})
+
+	s.Require().NoError(test.WaitForBlocks(ctx, 5, chainA, chainB), "failed to wait for blocks")
+
+	t.Run("Halt chain and export genesis", func(t *testing.T) {
+		s.HaltChainAndExportGenesis(ctx, chainA, int64(haltHeight))
+	})
+
+	t.Run("restart relayer", func(t *testing.T) {
+		s.StopRelayer(ctx, relayer)
+		s.StartRelayer(relayer)
+	})
+
+	t.Run("native IBC token transfer from chainA to chainB, sender is source of tokens", func(t *testing.T) {
+		transferTxResp, err := s.Transfer(ctx, chainA, chainAWallet, channelA.PortID, channelA.ChannelID, testvalues.DefaultTransferAmount(chainADenom), chainAAddress, chainBAddress, s.GetTimeoutHeight(ctx, chainB), 0, "")
+		s.Require().NoError(err)
+		s.AssertValidTxResponse(transferTxResp)
+	})
+
+	s.Require().NoError(test.WaitForBlocks(ctx, 5, chainA, chainB), "failed to wait for blocks")
+}
+
+func (s *GenesisTestSuite) HaltChainAndExportGenesis(ctx context.Context, chain *cosmos.CosmosChain, haltHeight int64) {
+	var genesisState GenesisState
+	
+	timeoutCtx, timeoutCtxCancel := context.WithTimeout(ctx, time.Minute*2)
+	defer timeoutCtxCancel()
+
+	height, err := chain.Height(ctx)
+	s.Require().NoError(err, "error fetching height before halt")
+	
+	err = test.WaitForBlocks(timeoutCtx, int(haltHeight - int64(height)) +1, chain)
+	s.Require().NoError(err)
+	err = chain.StopAllNodes(ctx)
+	s.Require().NoError(err, "error stopping node(s)")
+
+	state, err := chain.ExportState(ctx, int64(haltHeight-int64(height))+1)
+
+	s.Require().NoError(err)
+	err = tmjson.Unmarshal([]byte(state), &genesisState)
+	s.Require().NoError(err)
+	genesisJson, err := tmjson.MarshalIndent(genesisState, "", "  ")
+	s.Require().NoError(err)
+
+	err = WriteFile("genesis.json", genesisJson)
+	s.Require().NoError(err)
+
+	for _, node := range chain.FullNodes {
+		err = node.UnsafeResetAll(ctx)
+		s.Require().NoError(err)
+	}
+
+	// we are reinitializing the clients because we need to update the hostGRPCAddress after
+	// halt chain and subsequent restarting of nodes
+	s.InitGRPCClients(chain)
+
+	timeoutCtx, timeoutCtxCancel = context.WithTimeout(ctx, time.Minute*2)
+	defer timeoutCtxCancel()
+
+	err = test.WaitForBlocks(timeoutCtx, int(blocksAfterUpgrade), chain)
+	s.Require().NoError(err, "chain did not produce blocks after halt")
+
+	height, err = chain.Height(ctx)
+	s.Require().NoError(err, "error fetching height after halt")
+
+	s.Require().Greater(height, haltHeight, "height did not increment after halt")
+}
+
+func WriteFile(path string, body []byte) error {
```

Comment:
Hi @ThanhNhann, this function is going to write the contents of the genesis file to the host running `go test` not to the container where the chain is running, so the chains will not see this genesis file.

It looks like we don't have a function to copy a file **to** a container but [here](https://github.com/cosmos/ibc-go/blob/b4503f4b8f910d46e4c69fff8ea13e91812d8a75/e2e/dockerutil/dockerutil.go#L51) is a function we use to copy a file from a container.

You will likely need to write the contents of the file into all of the containers using the docker client in a manner similar to this.

Reply:
thank @chatton, it helps me a lot

Reply:
I will also add we want to make sure we don't re-import `tendermint` as we are now using `comeft`. In order to copy the the contents, it might make sense to just load the file as a `[]byte` using the function I linked and then write the file using a new function without needing to use `tmjson` at all.